tests/rptest/tests/datalake/datalake_verifier.py

```diff
@@ -0,0 +1,243 @@
+# Copyright 2024 Redpanda Data, Inc.
+#
+# Use of this software is governed by the Business Source License
+# included in the file licenses/BSL.md
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0
+
+from collections import defaultdict
+from concurrent.futures import ThreadPoolExecutor
+import random
+import threading
+import time
+from rptest.clients.rpk import RpkTool
+from rptest.services.redpanda import RedpandaService
+from rptest.tests.datalake.query_engine_base import QueryEngineBase
+from rptest.util import wait_until
+
+from confluent_kafka import Consumer
+
+
+class DatalakeVerifier():
+    """
+     Verifier that does the verification of the data in the redpanda Iceberg table. 
+     The verifier consumes offsets from specified topic and verifies it the data 
+     in the iceberg table matches.
+     
+     The verifier runs two threads: 
+     - one of them consumes messages from the specified topic and buffers them in memory. 
+       The semaphore is used to limit the number of messages buffered in memory.
+       
+     - second thread executes a per partition query that fetches the messages 
+       from the iceberg table
+    """
+
+    #TODO: add an ability to pass lambda to verify the message content
+    #TODO: add tolerance for compacted topics
+    def __init__(self, redpanda: RedpandaService, topic: str,
+                 query_engine: QueryEngineBase):
+        self.redpanda = redpanda
+        self.topic = topic
+        self.logger = redpanda.logger
+        self._consumed_messages = defaultdict(list)
+
+        self._query: QueryEngineBase = query_engine
+        self._cg = f"verifier-group-{random.randint(0, 1000000)}"
+        self._lock = threading.Lock()
+        self._offsets_lock = threading.Lock()
```

Comment:
nit: would benefit from a comment describing what each lock is protecting. Or maybe just have a single lock for simplicity.